src/ucc/api/ucc_def.h

```diff
@@ -63,6 +63,18 @@ typedef struct ucc_coll_req {
     ucc_status_t status;
 } ucc_coll_req_t;
 
+/**
+ * @ingroup UCC_COLLECTIVES_DT
+ * @brief UCC collective completion callback
+ *
+ * The callback is invoked whenever the collective operation is completed.
+ * It is not allowed to call UCC APIs from the completion callback except
+ * for @ref ucc_collective_finalize.
+ */
+typedef struct ucc_coll_callback {
+    void (*cb)(void *data, ucc_status_t status);
+    void  *data;
+} ucc_coll_callback_t;
```

Comment:
1. I guess the callback should be optional, right?
2. Does it have to be thread safe?

Reply:
1. It is optional (and passed on per collective basis, so user can decide for which colls to request the CB)
2. It is users responsibility to protect anything inside the  CB in the multi-threaded env.

Reply:
Makes sense. Let's put that in the document in the header.

Reply:
Another question - does mean every endpoint participating in the collectives should opt-in or opt-out? Can few endpoints decide to use callback and others decide to not use callback?

Reply:
As we discussed in WG: we don't put any restrictions with respect to that - up to user to not introduce dead locks.